File: sdk-node/tests/webhook/trust.test.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import type { EmailAuth, EmailReceivedEvent } from "../../src/types.js";
import { normalizeReceivedEmail } from "../../src/webhook/received-email.js";
import { isTrustedSender } from "../../src/webhook/trust.js";

const __dirname = dirname(fileURLToPath(import.meta.url));

const baseEvent = JSON.parse(
  readFileSync(
    join(__dirname, "../../../test-fixtures/webhook/valid-email-received.json"),
    "utf8",
  ),
) as EmailReceivedEvent;

function legitAuth(overrides: Partial<EmailAuth> = {}): EmailAuth {
  return {
    spf: "pass",
    dmarc: "pass",
    dmarcPolicy: "reject",
    dmarcFromDomain: "example.com",
    dmarcSpfAligned: true,
    dmarcDkimAligned: true,
    dmarcSpfStrict: false,
    dmarcDkimStrict: false,
    dkimSignatures: [
      {
        domain: "example.com",
        selector: "default",
        result: "pass",
        aligned: true,
        keyBits: 2048,
        algo: "rsa-sha256",
      },
    ],
    ...overrides,
  };
}

function buildEvent(overrides: {
  from?: string;
  mailFrom?: string;
  auth?: EmailAuth;
}): EmailReceivedEvent {
  const event = structuredClone(baseEvent);
  if (overrides.from !== undefined) {
    event.email.headers.from = overrides.from;
  }
  if (overrides.mailFrom !== undefined) {
    event.email.smtp.mail_from = overrides.mailFrom;
  }
  event.email.auth = overrides.auth ?? legitAuth();
  return event;
}

describe("isTrustedSender", () => {
  it("trusts authenticated mail from the expected domain", () => {
    const result = isTrustedSender(buildEvent({ from: "sender@example.com" }), {
      domain: "example.com",
    });
    expect(result).toMatchObject({
      trusted: true,
      retryable: false,
      reason: "trusted",
    });
    expect(result.auth.verdict).toBe("legit");
  });

  it("does not trust an unparseable From header even when the envelope sender matches", () => {
    // normalizeReceivedEmail falls back to smtp.mail_from when the From
    // header fails to parse; that fallback is attacker-controlled and must
    // never satisfy the trust check.
    const event = buildEvent({
      from: "not-an-email",
      mailFrom: "sender@example.com",
    });
    expect(normalizeReceivedEmail(event).sender.address).toBe(
      "sender@example.com",
    );
    const result = isTrustedSender(event, { domain: "example.com" });
    expect(result.trusted).toBe(false);
    expect(result.reason).toBe("from-header-invalid");
  });

  it("rejects a matching Reply-To when the From does not match", () => {
    const event = buildEvent({
      from: "attacker@evil.example.net",
      auth: legitAuth({
        dmarcFromDomain: "evil.example.net",
        dkimSignatures: [
          {
            domain: "evil.example.net",
            selector: "default",
            result: "pass",
            aligned: true,
            keyBits: 2048,
            algo: "rsa-sha256",
          },
        ],
      }),
    });
    event.email.parsed.reply_to = [
      { address: "sender@example.com", name: null },
    ];
    const result = isTrustedSender(event, { domain: "example.com" });
    expect(result.trusted).toBe(false);
    expect(result.reason).toBe("dmarc-domain-mismatch");
  });

  it("treats a missing auth object as untrusted instead of throwing", () => {
    const event = buildEvent({ from: "sender@example.com" });
    (event.email as { auth?: EmailAuth }).auth = undefined;
    const result = isTrustedSender(event, { domain: "example.com" });
    expect(result).toMatchObject({
      trusted: false,
      retryable: false,
      reason: "auth-missing",
    });
    expect(result.auth.verdict).toBe("unknown");
  });

  it("treats malformed dkimSignatures as untrusted instead of throwing", () => {
    const event = buildEvent({ from: "sender@example.com" });
    (
      event.email.auth as unknown as { dkimSignatures: unknown }
    ).dkimSignatures = "corrupt";
    const result = isTrustedSender(event, { domain: "example.com" });
    expect(result.reason).toBe("auth-missing");
  });

  it("rejects group syntax in the From header", () => {
    const result = isTrustedSender(
      buildEvent({ from: "Friends: sender@example.com;" }),
      { domain: "example.com" },
    );
    expect(result.trusted).toBe(false);
    expect(result.reason).toBe("from-header-invalid");
  });

  it("normalizes a whitespace-padded dmarcFromDomain", () => {
    const result = isTrustedSender(
      buildEvent({
        from: "sender@example.com",
        auth: legitAuth({ dmarcFromDomain: " Example.com " }),
      }),
      { domain: "example.com" },
    );
    expect(result.trusted).toBe(true);
  });

  it("does not treat an empty expected domain as matching an empty dmarcFromDomain", () => {
    expect(() =>
      isTrustedSender(buildEvent({ from: "sender@example.com" }), {
        domain: "  ",
      }),
    ).toThrow(TypeError);
  });

  it("throws TypeError for invalid options", () => {
    const event = buildEvent({ from: "sender@example.com" });
    expect(() =>
      isTrustedSender(event, { domain: "user@example.com" }),
    ).toThrow(TypeError);
    expect(() =>
      isTrustedSender(event, {
        domain: "example.com",
        sender: "not-an-email",
      }),
    ).toThrow(TypeError);
    expect(() =>
      isTrustedSender(event, {
        domain: "example.com",
        sender: "Name <sender@example.com>",
      }),
    ).toThrow(TypeError);
    expect(() => isTrustedSender(event, {} as { domain: string })).toThrow(
      TypeError,
    );
  });

  it("keeps the retryable flag false for every non-temperror reason", () => {
    const cases: Array<{ event: EmailReceivedEvent; domain: string }> = [
      {
        event: buildEvent({
          from: "sender@example.com",
          auth: legitAuth({
            spf: "fail",
            dmarc: "fail",
            dmarcSpfAligned: false,
            dmarcDkimAligned: false,
            dkimSignatures: [],
          }),
        }),
        domain: "example.com",
      },
      {
        event: buildEvent({
          from: "sender@example.com",
          auth: legitAuth({
            dmarc: "none",
            dmarcPolicy: null,
            dmarcSpfAligned: false,
            dmarcDkimAligned: false,
            dkimSignatures: [],
          }),
        }),
        domain: "example.com",
      },
      {
        event: buildEvent({ from: "sender@example.com" }),
        domain: "other.example.net",
      },
    ];
    for (const testCase of cases) {
      const result = isTrustedSender(testCase.event, {
        domain: testCase.domain,
      });
      expect(result.trusted).toBe(false);
      expect(result.retryable).toBe(false);
    }
  });
});

File: sdk-go/trust_test.go
package primitive

import (
	"strings"
	"testing"
)

func legitTrustAuth() EmailAuth {
	policy := DmarcPolicyReject
	return EmailAuth{
		SPF:              SpfResultPass,
		DMARC:            DmarcResultPass,
		DMARCPolicy:      &policy,
		DMARCFromDomain:  strPtr("example.com"),
		DMARCSpfAligned:  boolPtr(true),
		DMARCDkimAligned: boolPtr(true),
		DKIMSignatures: []DKIMSignature{{
			Domain:   "example.com",
			Selector: strPtr("default"),
			Result:   DkimResultPass,
			Aligned:  true,
			KeyBits:  intPtr(2048),
			Algo:     strPtr("rsa-sha256"),
		}},
	}
}

func trustEvent(from string, auth EmailAuth) EmailReceivedEvent {
	var event EmailReceivedEvent
	event.Email.Headers.From = from
	event.Email.Auth = auth
	return event
}

func TestIsTrustedSenderDomainMatch(t *testing.T) {
	result, err := IsTrustedSender(trustEvent("sender@example.com", legitTrustAuth()), TrustedSenderOptions{Domain: "example.com"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !result.Trusted || result.Retryable || result.Reason != TrustReasonTrusted {
		t.Fatalf("unexpected result: %#v", result)
	}
	if result.Auth.Verdict != AuthVerdictLegit {
		t.Fatalf("expected legit auth verdict, got %q", result.Auth.Verdict)
	}
}

func TestIsTrustedSenderExactSender(t *testing.T) {
	result, err := IsTrustedSender(trustEvent("sender@example.com", legitTrustAuth()), TrustedSenderOptions{Domain: "example.com", Sender: "SENDER@example.com"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !result.Trusted {
		t.Fatalf("expected trusted result: %#v", result)
	}

	result, err = IsTrustedSender(trustEvent("other@example.com", legitTrustAuth()), TrustedSenderOptions{Domain: "example.com", Sender: "sender@example.com"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Trusted || result.Reason != TrustReasonSenderMismatch {
		t.Fatalf("expected sender mismatch: %#v", result)
	}
}

func TestIsTrustedSenderWhitespacePaddedDmarcDomain(t *testing.T) {
	auth := legitTrustAuth()
	padded := " Example.com "
	auth.DMARCFromDomain = &padded
	result, err := IsTrustedSender(trustEvent("sender@example.com", auth), TrustedSenderOptions{Domain: "example.com"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !result.Trusted {
		t.Fatalf("expected trusted result: %#v", result)
	}
}

func TestIsTrustedSenderZeroValueAuth(t *testing.T) {
	result, err := IsTrustedSender(trustEvent("sender@example.com", EmailAuth{}), TrustedSenderOptions{Domain: "example.com"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Trusted || result.Retryable {
		t.Fatalf("expected permanent untrusted result: %#v", result)
	}
}

func TestIsTrustedSenderInvalidOptions(t *testing.T) {
	event := trustEvent("sender@example.com", legitTrustAuth())
	for name, opts := range map[string]TrustedSenderOptions{
		"empty domain":         {Domain: "   "},
		"address as domain":    {Domain: "user@example.com"},
		"invalid sender":       {Domain: "example.com", Sender: "not-an-email"},
		"display-name sender":  {Domain: "example.com", Sender: "Name <sender@example.com>"},
		"whitespace in domain": {Domain: "exam ple.com"},
	} {
		if _, err := IsTrustedSender(event, opts); err == nil {
			t.Fatalf("%s: expected error", name)
		}
	}
}

func TestParseFromHeaderStrictGroupSyntax(t *testing.T) {
	// mail.ParseAddressList flattens a single-member group to its
	// member, so the explicit group-delimiter guard must reject it to
	// stay aligned with the Node parser.
	if _, reason := parseFromHeaderStrict("Friends: sender@example.com;"); reason != TrustReasonFromHeaderInvalid {
		t.Fatalf("expected group-syntax rejection, got %q", reason)
	}
	// A colon inside a quoted display name is not group syntax.
	if address, _ := parseFromHeaderStrict(`"Support: Billing" <sender@example.com>`); address != "sender@example.com" {
		t.Fatalf("expected quoted-colon display name to parse, got %q", address)
	}
	// An escaped quote inside a quoted display name does not end the
	// quoted string; the colon after it stays quoted.
	if address, _ := parseFromHeaderStrict(`"a\": b" <sender@example.com>`); address != "sender@example.com" {
		t.Fatalf("expected escaped-quote display name to parse, got %q", address)
	}
	// A colon inside an IPv6 domain literal is rejected, matching the
	// Node parser (whose lexer also treats it as a group marker).
	if _, reason := parseFromHeaderStrict("user@[IPv6:2001:db8::1]"); reason != TrustReasonFromHeaderInvalid {
		t.Fatalf("expected IPv6 domain-literal rejection, got %q", reason)
	}
}

func TestParseFromHeaderStrict(t *testing.T) {
	if address, _ := parseFromHeaderStrict(`"spoofed@example.com" <attacker@evil.example.net>`); address != "attacker@evil.example.net" {
		t.Fatalf("expected angle-addr address, got %q", address)
	}
	if _, reason := parseFromHeaderStrict("a@example.com, b@example.com"); reason != TrustReasonFromHeaderMultipleAddresses {
		t.Fatalf("expected multiple-address rejection, got %q", reason)
	}
	if _, reason := parseFromHeaderStrict("not-an-email"); reason != TrustReasonFromHeaderInvalid {
		t.Fatalf("expected invalid rejection, got %q", reason)
	}
	if _, reason := parseFromHeaderStrict(""); reason != TrustReasonFromHeaderInvalid {
		t.Fatalf("expected empty rejection, got %q", reason)
	}
	if _, reason := parseFromHeaderStrict("a@" + strings.Repeat("x", maxHeaderBytes) + ".com"); reason != TrustReasonFromHeaderInvalid {
		t.Fatalf("expected over-length rejection, got %q", reason)
	}
}
